### PR TITLE
Merge upstream

### PR DESCRIPTION
**Pull request by sadamov**
This PR consists of a merge with upstream/main including reformating and linting using new pre-comits. I suggest to review this PR by doing the following (instead of checking each line change):

- Create env with pip
- Create env with conda
- Run `slurm_train.sh`
- Run `slurm_eval.sh`

If all runs through smoothly without errors, we can merge. I am handling this a bit like an initial commit.

### CODE REVIEW

**Comment by sadamov** [comment]
@joeloskarsson Just wanted to let you know that I have prepared myself for our co-coding session in May :computer:. Our main-branch is up-to-date with yours :handshake:  Thanks again for implementing the pre-commits!

File: .vscode/launch.json
Comment (twicki) [inline .vscode/launch.json:1]: I think your entire `.vscode` folder should not be committed.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    "__init__.py: F401", # Allow unused imports
+]
+
+[tool.codespell]
+skip = "requirements/*"
+
+# Pylint config
+[tool.pylint]
+ignore = [
+    "create_mesh.py", # Disable linting for now, as major rework is planned/expected
```
Comment (twicki) [inline pyproject.toml:42]: do we need to open an issue for that to keep it in mind?

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+shapely>=2.0.1
+networkx>=3.0
+imageio>=2.34.0
+numcodecs>=0.12.1
+Cartopy>=0.22.0
+pyproj>=3.4.1
+xarray>=2024.1.1
+tueplots>=0.0.8
+plotly>=5.15.0
+# for dev
```
Comment (twicki) [inline requirements.txt:16]: normally you would do a two-stepped approach here where you specify the two requirements in two files. See here for example:
https://github.com/MeteoSwiss-APN/nwp_log_mining/blob/main/requirements/requirements_build.txt
https://github.com/MeteoSwiss-APN/nwp_log_mining/blob/main/requirements/requirements_run.txt

File: slurm_eval.sh [hunk 3]
```diff
@@ -16,18 +16,21 @@ export NORMALIZE=false
...
         srun -ul -N1 -n1 python create_parameter_weights.py --dataset "cosmo" --batch_size 32 --n_workers 8 --step_length 1
     fi
 fi
 
 ulimit -c 0
 export OMP_NUM_THREADS=16
 
-# Run the script with torchrun
 srun -ul python train_model.py --load "wandb/example.ckpt" --dataset "cosmo" \
-    --eval="test" --subset_ds 1 --n_workers 2 --batch_size 6 --wandb_mode "offline"
```
Comment (twicki) [inline slurm_eval.sh:33]: why can we switch off offline now?

File: train_model.py [hunk 4]
```diff
@@ -1,14 +1,17 @@
+# pylint: disable=wrong-import-order
```
Comment (twicki) [inline train_model.py:1]: why is that needed?

**Review by twicki — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (sadamov) [reply pyproject.toml:42]: The issue already exists upstream: https://github.com/joeloskarsson/neural-lam/issues/4

File: requirements.txt (hunk 2 above)
Comment (sadamov) [reply requirements.txt:16]: This is directly defined by upstream. If you have strong feelings about it I can create a PR? otherwise I would suggest to use the conda within MeteoSwiss as it is easier and consistent with our common practices.

File: slurm_eval.sh (hunk 3 above)
Comment (sadamov) [reply slurm_eval.sh:33]: As shown in the `README.md` it is preferred to give the option to the user whether to `wandb login` or `wandb off`. I am not 100% sure how this setting interacts with slurm/sbatch scripts though.

File: train_model.py (hunk 4 above)
Comment (sadamov) [reply train_model.py:1]: It isn't :D I was having issues with my local ruff-formatter. Fixed in the latest commit.

Comment (sadamov) [reply train_model.py:1]: Aha, I just realized that pylint and isort do not agree on the order either:
`train_model.py:10:0: C0411: third party import "from lightning_fabric.utilities import seed" should be placed before "import wandb" (wrong-import-order)` 
If it is okay with you, I would ignore the message for now.

Comment (twicki) [reply train_model.py:1]: I fixed this in a PR to this branch
